# Практическая работа 4. Данные в формате CSV

###Выполнил студент ПИН-Б-З-22-1 Гадиян Сергей Гариевич<br>
##Индивидуальное задание находится в конце

## Задание

В data_v3.csv присутствуют следующие поля: `ID` – уникальный ключ записи; `LastName`, `firstName` – фамилия и имя держателя карты; `email` – электронная почта; `Plus` и `Minus` – суммы поступлений и списаний по карте соответственно в долларах; `credit_card` – номер карты; `Tax` – уплаченный налог.
Необходимо ответить на следующий вопросы:
1. Найдите держателя карты с максимальной прибылью. В ответе необходимо указать ФИО держателя и максимальную прибыль (прибыль = поступления – списания).
2. Найдите количество держателей карт, у которых сумма уплаченного налога больше среднего арифметического значения налога по всей выборке.


## Анализ задания, получение данных

Файл data_v3.txt расположен репозитории. Выполним копирование файла в локальное хранилище Google Colab, используя следующую команду:

In [2]:
!wget https://raw.githubusercontent.com/enikolaev/IT-and-Programming/main/data/data_v3.csv


--2025-06-10 20:52:22--  https://raw.githubusercontent.com/enikolaev/IT-and-Programming/main/data/data_v3.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 92780 (91K) [text/plain]
Saving to: ‘data_v3.csv’

data_v3.csv         100%[===================>]  90.61K  --.-KB/s    in 0.02s   

2025-06-10 20:52:22 (3.68 MB/s) - ‘data_v3.csv’ saved [92780/92780]



Можно скопировать и поместить файл с данными любым удобным способом

Реализуем чтение всего файла построчно. Используем метод `open()`, а затем будем обходить файл как коллекцию с использованием цикла `for line in File`. Очевидно, что csv-файл используется в данной работе как обычный текстовый файл и в цикле `for` производится последовательный обход всех строк этого файла, то есть текстовый файл - это коллекция его строк.
В файле много строк, поэтому будем выводить только первые `n_to_print` строк, но обрабатывать будем все (чтобы не возникло скрытых ошибок)

In [ ]:
F=open('data_v3.csv')
n_to_print=10
for line in F:
  n_to_print-=1
  if n_to_print>0:
    print(line)

ID,LastName,FirstName,email,Plus,Minus,credit_card,Tax

12-9321320,Pickle,Réjane,,$41564.80,$46160.22,3557725016917448,$16408.20

14-4213773,Hartford,Léa,,$46310.29,$13116.21,3544948067779642,$13397.19

92-0579703,Caldecott,Adélaïde,lcaldecott2@com.com,$44330.64,$9725.39,4508863396652307,$35768.27

43-3066209,Dohrmann,Cloé,odohrmann3@gov.uk,$32271.10,$33467.96,6398709109212517,$3876.95

39-1753452,Pes,Zhì,dpes4@trellian.com,$24467.01,$5220.55,3562917312487449,$5559.11

98-1640163,Satyford,Pénélope,asatyford5@friendfeed.com,$31883.05,$11529.92,3564171803659243,$5939.26

58-9143184,Corballis,Clémentine,scorballis6@mapquest.com,$25340.90,$3241.86,341075109903237,$45625.23

87-3594410,Hiseman,Bénédicte,chiseman7@sciencedirect.com,$46081.76,$36954.21,3548420880126084,$19464.18



Строки перебираются, но теперь необходимо разделить единую строку на отдельные элементы данных. Для этого выполним метод `split(sep)`, который разрезает строку, используя символ `sep` в качестве разделителя. Также обратим внимание на тот факт, что в первой строке содержатся не данные, а заголовки столбцов.

In [ ]:
# доработка предыдущего кода
F=open('data_v3.csv')
header = F.readline().strip().split(',')
print(header)
print('-'*20)
n_to_print=10
for line in F:
  # производим обработку в любом случае ...
  data = line.strip().split(',')
  # ... аа печатаем только первые n_to_print
  n_to_print-=1
  if n_to_print>0:
    print(data)

['ID', 'LastName', 'FirstName', 'email', 'Plus', 'Minus', 'credit_card', 'Tax']
--------------------
['12-9321320', 'Pickle', 'Réjane', '', '$41564.80', '$46160.22', '3557725016917448', '$16408.20']
['14-4213773', 'Hartford', 'Léa', '', '$46310.29', '$13116.21', '3544948067779642', '$13397.19']
['92-0579703', 'Caldecott', 'Adélaïde', 'lcaldecott2@com.com', '$44330.64', '$9725.39', '4508863396652307', '$35768.27']
['43-3066209', 'Dohrmann', 'Cloé', 'odohrmann3@gov.uk', '$32271.10', '$33467.96', '6398709109212517', '$3876.95']
['39-1753452', 'Pes', 'Zhì', 'dpes4@trellian.com', '$24467.01', '$5220.55', '3562917312487449', '$5559.11']
['98-1640163', 'Satyford', 'Pénélope', 'asatyford5@friendfeed.com', '$31883.05', '$11529.92', '3564171803659243', '$5939.26']
['58-9143184', 'Corballis', 'Clémentine', 'scorballis6@mapquest.com', '$25340.90', '$3241.86', '341075109903237', '$45625.23']
['87-3594410', 'Hiseman', 'Bénédicte', 'chiseman7@sciencedirect.com', '$46081.76', '$36954.21', '35484208801

Метод `split()` вернул список отдельных строк - полстрок исходной строки. Теперь необходимо сформировать структуру данных максимально полно описывающую предметную область.
Для этого сформируем список словарей, то есть все данные в процессе чтения будем построчно перемещать в список `all_clients`, при этом каждый клиент будет представлен в виде словаря. Например первая строка в файле будет выглядеть следующим образом:

`example = {'ID' : '12-9321320', 'LastName':'Pickle', 'FirstName':'Réjane', 'email':None, 'Plus':41564.80, 'Minus':46160.22, 'credit_card':'3557725016917448', 'Tax':16408.20}`

то есть будет сформирована структура данных в которой все элементыы представляют собой пары вида `ключ` : `значение`, при этом как и в списке возможно обращение к элементамм данной структуры по индексу, только в словарях вместо целочисленного индекса используется ключ.
Например, применительно к первой строке, обратиться к сумме поступления для ее печати можно следующим образом:

`print( example['Plus'] )`

In [ ]:
# Доработаем предыдущий скрипт
F=open('data_v3.csv')
header = F.readline().strip().split(',')

# список всех клиентов
all_clients=[]

for line in F:
  data = line.strip().split(',')
  client = {} # создаем пустой словарь
  for i, key in enumerate(header): # обходим все заголовки
    client[key] = data[i] # добавляем в словарь клюс:значение

  # некоторые значения придется конвертировать к подходящим типам
  client['Minus'] = float(client['Minus'][1:])
  client['Plus'] = float(client['Plus'][1:])
  client['Tax'] = float(client['Tax'][1:])

  all_clients.append(client) # добавляем клиента в скписок

print(all_clients[:10]) # напечатаем первых 10 клиентов

F.close() # файл больше не нужен

[{'ID': '12-9321320', 'LastName': 'Pickle', 'FirstName': 'Réjane', 'email': '', 'Plus': 41564.8, 'Minus': 46160.22, 'credit_card': '3557725016917448', 'Tax': 16408.2}, {'ID': '14-4213773', 'LastName': 'Hartford', 'FirstName': 'Léa', 'email': '', 'Plus': 46310.29, 'Minus': 13116.21, 'credit_card': '3544948067779642', 'Tax': 13397.19}, {'ID': '92-0579703', 'LastName': 'Caldecott', 'FirstName': 'Adélaïde', 'email': 'lcaldecott2@com.com', 'Plus': 44330.64, 'Minus': 9725.39, 'credit_card': '4508863396652307', 'Tax': 35768.27}, {'ID': '43-3066209', 'LastName': 'Dohrmann', 'FirstName': 'Cloé', 'email': 'odohrmann3@gov.uk', 'Plus': 32271.1, 'Minus': 33467.96, 'credit_card': '6398709109212517', 'Tax': 3876.95}, {'ID': '39-1753452', 'LastName': 'Pes', 'FirstName': 'Zhì', 'email': 'dpes4@trellian.com', 'Plus': 24467.01, 'Minus': 5220.55, 'credit_card': '3562917312487449', 'Tax': 5559.11}, {'ID': '98-1640163', 'LastName': 'Satyford', 'FirstName': 'Pénélope', 'email': 'asatyford5@friendfeed.com', '

## Решение задачи 1

Первая задача:
Найдите держателя карты с максимальной прибылью. В ответе необходимо указать ФИО держателя и максимальную прибыль (прибыль = поступления – списания).

Переберем все элементы в списке `all_clients` и выберем клиента с максимальной разницей `client['Plus'] - client['Minus']`



In [ ]:
# переменные для хранения целевого клиента и максимальной прибыли
target_client=[]
max_profit=0

for client in all_clients:
  if client['Plus'] - client['Minus'] > max_profit:
    max_profit = client['Plus'] - client['Minus']
    target_client = client

print('Максимальная прибыль: ', max_profit)
print('Клиент с максимальной прибылью', target_client)

Максимальная прибыль:  45913.590000000004
Клиент с максимальной прибылью {'ID': '98-9398689', 'LastName': 'Lingwood', 'FirstName': 'Maïté', 'email': 'mlingwood1y@wordpress.com', 'Plus': 47412.86, 'Minus': 1499.27, 'credit_card': '201561279094179', 'Tax': 23816.11}


## Решение задачи 2

Вторая задача: Найдите количество держателей карт, у которых сумма уплаченного налога больше среднего арифметического значения налога по всей выборке.

Задачу можно решить также перебором всей выборки, но предварительно необходимо определить среднее значение уплаченных налогов

In [ ]:
# Определим среднее значение нолога
total_tax = 0
count_tax = len(all_clients)

for client in all_clients:
  total_tax += client['Tax']

avg_tax = total_tax / count_tax

print('Среднее значение налога = %.2f' % avg_tax)

counter=0
for client in all_clients:
  if client['Tax'] > avg_tax:
    counter+=1

print('Количество клиентов с налогом выше среднего: ', counter)

Среднее значение налога = 24939.93
Количество клиентов с налогом выше среднего:  489


В данной работе были использованы операторы `if` (условный оператор), `for ... in range(...)` (оператор цикла с обходом диапазона). Применялись функции `print()`, `len()`, `max()`. В работе использовался тип данных `dict()`

##Индивидуальное задание

Вариант 4
1. Количество сотрудников в каждой из 4-х компаний;
2. Среднюю з/п в каждой из 4-х компаний;
3. Количество сотрудников, владеющих английским языком;
4. Количество сотрудников, получающих более $3000,00.

Получим файл с нашим вариантом (4).

In [14]:
!wget https://raw.githubusercontent.com/enikolaev/IT-and-Programming/main/data/pract7/p7_data_04.csv

--2025-06-10 20:53:14--  https://raw.githubusercontent.com/enikolaev/IT-and-Programming/main/data/pract7/p7_data_04.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 53585 (52K) [text/plain]
Saving to: ‘p7_data_04.csv.1’

p7_data_04.csv.1    100%[===================>]  52.33K  --.-KB/s    in 0.01s   

2025-06-10 20:53:14 (4.11 MB/s) - ‘p7_data_04.csv.1’ saved [53585/53585]



In [16]:
import csv  # импорт модуля для работы с CSV-файлами
from collections import defaultdict  # импорт defaultdict для удобной работы со словарями
# чтение данных из CSV-файла
with open('p7_data_04.csv', 'r') as f:  # открываем файл для чтения
  reader = csv.DictReader(f)  # создаем объект DictReader для чтения данных как словарей
  data = list(reader)  # преобразуем данные в список словарей для удобной обработки
#  Задача 1:Количество сотрудников в каждой компании
company_counts = defaultdict(int)  # создаем словарь с автоматической инициализацией int
for row in data:  # перебираем каждую строку данных
  company_counts[row['Company']] += 1  # увеличиваем счетчик для текущей компании
print("Количество сотрудников в каждой компании:")  # выводим заголовок
for company, count in company_counts.items():  # перебираем компании и их количество
  print(f"{company}: {count}")  # выводим результат
# Задание 2 Средняя зарплата в каждой компании
company_salaries = defaultdict(list)  # словарь для хранения зарплат по компаниям
for row in data:  # перебираем строки данных
  salary = float(row['Salary'].replace('$', '').replace(',', ''))  # преобразуем зарплату в число
company_salaries[row['Company']].append(salary)  # добавляем зарплату в список для компании
print("\nСредняя зарплата в каждой компании:")  # выводим заголовок
for company, salaries in company_salaries.items():  # перебираем компании и их зарплаты
  avg_salary = sum(salaries) / len(salaries)  # вычисляем среднюю зарплату
print(f"{company}: ${avg_salary:.2f}")  # выводим результат с 2 знаками после запятой
# Задание 3 Количество сотрудников, владеющих английским языком
english_speakers = 0  # инициализируем счетчик
for row in data:  # перебираем строки данных
  if 'English' in row['Language']:  # проверяем наличие английского в языках
    english_speakers += 1  # увеличиваем счетчик
print(f"\nКоличество сотрудников, владеющих английским: {english_speakers}")  # Выводим результат
# Задание 4 Количество сотрудников с зарплатой более $3000
high_salary_employees = 0  # инициализируем счетчик
for row in data:  # перебираем строки данных
  salary = float(row['Salary'].replace('$', '').replace(',', ''))  # преобразуем зарплату в число
if salary > 3000:  # проверяем условие
  high_salary_employees += 1  # увеличиваем счетчик
print(f"\nКоличество сотрудников с зарплатой более $3000: {high_salary_employees}")  # выводим результат

Количество сотрудников в каждой компании:
Zuko Ltd: 255
Workus Co: 257
Craft Ltd: 242
NodaTLT Co: 246

Средняя зарплата в каждой компании:
Craft Ltd: $1392.62

Количество сотрудников, владеющих английским: 250

Количество сотрудников с зарплатой более $3000: 0


###Ответы на контрольные вопросы

1. Напишите код для печати суммы элементов списка p: <br>
p = [10, 100, 34, 56, 45, 45]


In [ ]:
p = [10, 100, 34, 56, 45, 45]
print("Сумма элементов:", sum(p))

Сумма элементов: 290


2. Напишите код для печати количества элементов списка p: <Br>
p = [10, 100, 34, 56, 45, 45]

In [ ]:
p = [10, 100, 34, 56, 45, 45]
print("Количество элементов:", len(p))

Количество элементов: 6


3. Напишите код для печати суммы элементов кратных 7: <br>
p = [10, 100, 35, 34, 56, 45, 45, 21]

In [ ]:
p = [10, 100, 35, 34, 56, 45, 45, 21]
sum_7 = sum(x for x in p if x % 7 == 0)
print("Сумма элементов, кратных 7:", sum_7)

Сумма элементов, кратных 7: 112


4. Напишите код для поиска ключей словаря d, для которых значения
превышают 45: <br>
p = {4:10, 12:100, 35:3, 34:56, 56:11, 45:’abc’}

In [ ]:
d = {4: 10, 12: 100, 35: 3, 34: 56, 56: 11, 45: 'abc'}
keys = [k for k, v in d.items() if isinstance(v, int) and v > 45]
print("Ключи с значениями > 45:", keys)

Ключи с значениями > 45: [12, 34]


5. Напишите код, который выводит список, составленный из квадратов чисел
от 1 до 100.<br> Сам список также необходимо сгенерировать.


In [ ]:
squares = [i**2 for i in range(1, 101)]
print("Квадраты чисел от 1 до 100:", squares)

Квадраты чисел от 1 до 100: [1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169, 196, 225, 256, 289, 324, 361, 400, 441, 484, 529, 576, 625, 676, 729, 784, 841, 900, 961, 1024, 1089, 1156, 1225, 1296, 1369, 1444, 1521, 1600, 1681, 1764, 1849, 1936, 2025, 2116, 2209, 2304, 2401, 2500, 2601, 2704, 2809, 2916, 3025, 3136, 3249, 3364, 3481, 3600, 3721, 3844, 3969, 4096, 4225, 4356, 4489, 4624, 4761, 4900, 5041, 5184, 5329, 5476, 5625, 5776, 5929, 6084, 6241, 6400, 6561, 6724, 6889, 7056, 7225, 7396, 7569, 7744, 7921, 8100, 8281, 8464, 8649, 8836, 9025, 9216, 9409, 9604, 9801, 10000]


6. Опишите структуру файла csv. Поясните последовательность действий при
чтении файла csv.

Структура CSV:


Строки разделены символами \n.

Столбцы в строке разделены запятыми (,) или другим разделителем.

Первая строка обычно содержит заголовки.


```
import csv

with open('file.csv', 'r') as f:
    reader = csv.reader(f)  # Или csv.DictReader для словарей
    headers = next(reader)  # Заголовки (если есть)
    for row in reader:
        print(row)  # Обработка данных
```
- Открытие файла: Используется with open(...) для автоматического закрытия.

- Чтение данных: csv.reader читает строки как списки, csv.DictReader — как словари с заголовками.

- Обработка: Данные обрабатываются построчно.
